src/prommis/nanofiltration/membrane_cascade_flowsheet/diafiltration_flowsheet_model.py

```diff
@@ -0,0 +1,1120 @@
+                                 * pyo.units.kg / pyo.units.hour)
+            )
+
+        @m.Expression()
+        def rec_perc_li(b):
+            return (
+                b.rec_mass_li / ((self.feed['Li']+self.diaf['Li'])
+                                 * pyo.units.kg / pyo.units.hour)
+            )
+
+        # mass recovery objectives
+        m.co_obj = pyo.Objective(
+            expr=m.rec_mass_co,
+            sense=pyo.maximize
+        )
+
+        m.li_obj = pyo.Objective(
+            expr=m.rec_mass_li,
+            sense=pyo.maximize
+        )
+        m.li_obj.deactivate()    # normally using Co as objective
+
+        # percent recovery lower bound constraints
+        m.R = pyo.Param(initialize=0.95, mutable=True)
+
+        @m.Constraint()
+        def li_lb(b):
+            return b.rec_perc_li >= m.R
+
+        @m.Constraint()
+        def co_lb(b):
+            return b.rec_perc_co >= m.R
+        m.co_lb.deactivate()    # normally use Co as objective
+
+        # product stream purities
+        @m.Expression()
+        def purity_co(b):
+            return (
+                sum(
+                    m.fs.split_retentate[i].product.mass_solute[0, 'Co']
+                    for i in pyo.RangeSet(self.ns)
+                ) / sum(
+                    m.fs.split_retentate[i].product.mass_solute[0, j]
+                    for i in pyo.RangeSet(self.ns)
+                    for j in self.solutes
+                )
+            )
+
+        @m.Expression()
+        def purity_li(b):
+            return (
+                sum(
+                    m.fs.split_permeate[i].product.mass_solute[0, 'Li']
+                    for i in pyo.RangeSet(self.ns)
+                ) / sum(
+                    m.fs.split_permeate[i].product.mass_solute[0, j]
+                    for i in pyo.RangeSet(self.ns)
+                    for j in self.solutes
+                )
+            )
+
+        m.pure = pyo.Param(initialize=0.1, mutable=True)
+
+        @m.Constraint()
+        def purity_co_lb(b):
+            return (
+                sum(
+                    m.fs.split_retentate[i].product.mass_solute[0, 'Co']
+                    for i in pyo.RangeSet(self.ns)
+                ) >= m.pure * sum(
+                    m.fs.split_retentate[i].product.mass_solute[0, j]
+                    for i in pyo.RangeSet(self.ns)
+                    for j in self.solutes
+                )
+            )
+
+        @m.Constraint()
+        def purity_li_lb(b):
+            return (
+                sum(
+                    m.fs.split_permeate[i].product.mass_solute[0, 'Li']
+                    for i in pyo.RangeSet(self.ns)
+                ) >= m.pure * sum(
+                    m.fs.split_permeate[i].product.mass_solute[0, j]
+                    for i in pyo.RangeSet(self.ns)
+                    for j in self.solutes
+                )
+            )
+        m.purity_co_lb.deactivate()
+        m.purity_li_lb.deactivate()
```

Comment:
Should one of these constraints (probably the Co one) be activated?

Reply:
The purity constraints are provided but optional. I have them deactivated normally and only activate them when needed after the flowsheet has already been constructed.